pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -140,41 +141,42 @@ type reconciler struct {
 	volumePluginMgr               *volumepkg.VolumePluginMgr
 	kubeletPodsDir                string
 	timeOfLastSync                time.Time
+	possiblyOrphanVolumes         []podVolume
 }
 
 func (rc *reconciler) Run(stopCh <-chan struct{}) {
-	wait.Until(rc.reconciliationLoopFunc(), rc.loopSleepDuration, stopCh)
-}
-
-func (rc *reconciler) reconciliationLoopFunc() func() {
-	return func() {
-		rc.reconcile()
-
-		// Sync the state with the reality once after all existing pods are added to the desired state from all sources.
-		// Otherwise, the reconstruct process may clean up pods' volumes that are still in use because
-		// desired state of world does not contain a complete list of pods.
-		if rc.populatorHasAddedPods() && !rc.StatesHasBeenSynced() {
-			klog.InfoS("Reconciler: start to sync state")
-			rc.sync()
-		}
-	}
+	rc.reconstructVolumes()
+	klog.InfoS("Reconciler: start to sync state")
+	wait.Until(rc.reconcile, rc.loopSleepDuration, stopCh)
 }
 
 func (rc *reconciler) reconcile() {
-	// Unmounts are triggered before mounts so that a volume that was
-	// referenced by a pod that was deleted and is now referenced by another
-	// pod is unmounted from the first pod before being mounted to the new
-	// pod.
-	rc.unmountVolumes()
+	dswPopulated := rc.populatorHasAddedPods()
+	// Unmount volumes only when DSW is fully populated to prevent unmounting a volume
+	// that is still needed, but it did not reach DSW yet.
+	if dswPopulated {
+		// Unmounts are triggered before mounts so that a volume that was
+		// referenced by a pod that was deleted and is now referenced by another
+		// pod is unmounted from the first pod before being mounted to the new
+		// pod.
+		rc.unmountVolumes()
+	}
 
 	// Next we mount required volumes. This function could also trigger
 	// attach if kubelet is responsible for attaching volumes.
 	// If underlying PVC was resized while in-use then this function also handles volume
 	// resizing.
 	rc.mountAttachVolumes()
 
-	// Ensure devices that should be detached/unmounted are detached/unmounted.
-	rc.unmountDetachDevices()
+	// Unmount volumes only when DSW is fully populated to prevent unmounting a volume
+	// that is still needed, but it did not reach DSW yet.
+	if dswPopulated {
```

Comment:
actually this block can be combined with line 157. I understand this ordering was there before, not sure the exact reason for it.